### PR TITLE
ARUHA-349 Recreate producer in specific cases

### PR DESCRIPTION
**Pull request by antban**

### CODE REVIEW

File: src/main/java/org/zalando/nakadi/repository/kafka/KafkaFactory.java [hunk 1]
```diff
@@ -8,22 +15,73 @@
...
+        } finally {
+            rwLock.readLock().unlock();
+        }
+        rwLock.writeLock().lock();
+        try {
+            if (null != activeProducer) {
+                useCount.get(activeProducer).incrementAndGet();
+                return activeProducer;
+            }
+            activeProducer = new KafkaProducer<>(kafkaLocationManager.getKafkaProducerProperties());
```
Comment (rcillo) [inline src/main/java/org/zalando/nakadi/repository/kafka/KafkaFactory.java]: It would make life easier if we logged something when a new broker is created and also when it's deleted in line 70.

File: src/main/java/org/zalando/nakadi/domain/BatchItem.java [hunk 2]
```diff
@@ -29,18 +28,22 @@ public void setPartition(final String partition) {
 
     @Nullable
     public String getPartition() {
-        return this.partition;
+        return partition;
```
Comment (adyach) [inline src/main/java/org/zalando/nakadi/domain/BatchItem.java]: Why did you remove 'this' here but left it for other methods?

**Comment by v-stepanov** [comment]
I wanted to take a look, so I will put a block for now :)

**Comment by v-stepanov** [comment]
👎

**Comment by v-stepanov** [comment]
I think there should be tests for new functionality

File: src/main/java/org/zalando/nakadi/repository/kafka/KafkaFactory.java [hunk 3]
```diff
@@ -8,22 +15,73 @@
...
+            useCount.put(activeProducer, new AtomicInteger(1));
+            return activeProducer;
+        } finally {
+            rwLock.writeLock().unlock();
+        }
+    }
+
+    public void releaseProducer(final Producer<String, String> producer) {
+        final int newUseCount = useCount.get(producer).decrementAndGet();
+        assert newUseCount >= 0;
```
Comment (adyach) [inline src/main/java/org/zalando/nakadi/repository/kafka/KafkaFactory.java]: Usually, assertions are used during development. Maybe you will remove it or replace with throwing exception?

Comment (antban) [inline src/main/java/org/zalando/nakadi/repository/kafka/KafkaFactory.java]: I wll just use different if condition

File: src/main/java/org/zalando/nakadi/repository/kafka/KafkaFactory.java (hunk 1 above)
Comment (antban) [inline src/main/java/org/zalando/nakadi/repository/kafka/KafkaFactory.java]: Makes sence

File: src/main/java/org/zalando/nakadi/repository/kafka/KafkaFactory.java [hunk 4]
```diff
@@ -8,22 +15,73 @@
...
-
 @Component
 @Profile("!test")
 public class KafkaFactory {
+
     private final KafkaLocationManager kafkaLocationManager;
-    private final KafkaProducer<String, String> kafkaProducer;
+    @Nullable
+    private Producer<String, String> activeProducer;
+    private final Map<Producer<String, String>, AtomicInteger> useCount = new ConcurrentHashMap<>();
```
Comment (adyach) [inline src/main/java/org/zalando/nakadi/repository/kafka/KafkaFactory.java]: AFAIU, it is possible to replace Map with AtomicInteger because you use only one producer all the time. What do you think ?

File: src/main/java/org/zalando/nakadi/repository/kafka/KafkaFactory.java [hunk 5]
```diff
@@ -8,22 +15,73 @@
...
+            activeProducer = new KafkaProducer<>(kafkaLocationManager.getKafkaProducerProperties());
+            useCount.put(activeProducer, new AtomicInteger(1));
+            return activeProducer;
+        } finally {
+            rwLock.writeLock().unlock();
+        }
+    }
+
+    public void releaseProducer(final Producer<String, String> producer) {
+        final int newUseCount = useCount.get(producer).decrementAndGet();
```
Comment (adyach) [inline src/main/java/org/zalando/nakadi/repository/kafka/KafkaFactory.java]: This value can inconsistence in a way if `takeProducer` is be called after `releaseProducer`.

File: src/main/java/org/zalando/nakadi/repository/kafka/KafkaTopicRepository.java [hunk 6]
```diff
@@ -164,43 +165,89 @@ public boolean partitionExists(final String topic, final String partition) throw
...
-                    .anyMatch(item -> item.getResponse().getPublishingStatus() == EventPublishingStatus.FAILED);
-            if (atLeastOneFailed) {
-                failUnpublished(batch, "internal error");
-                throw new EventPublishingException("Error publishing message to kafka");
+            final CompletableFuture<Void> multiFuture = CompletableFuture.allOf(
+                    sendFutures.values().toArray(new CompletableFuture<?>[sendFutures.size()]));
+            multiFuture.get(createSendTimeout(), TimeUnit.MILLISECONDS);
+
+            // Now lets check for errors
+            final boolean needReset = sendFutures.entrySet().stream().filter(
```
Comment (adyach) [inline src/main/java/org/zalando/nakadi/repository/kafka/KafkaTopicRepository.java]: For me this one is cleaner:
            `sendFutures.entrySet().stream().filter(
                    entry -> isExceptionShouldLeadToReset(entry.getValue().getNow(null)))
                    .findAny()
                    .ifPresent(entry -> kafkaFactory.terminateProducer(producer));`

File: src/main/java/org/zalando/nakadi/repository/kafka/KafkaTopicRepository.java [hunk 7]
```diff
@@ -164,43 +165,89 @@ public boolean partitionExists(final String topic, final String partition) throw
...
+            item.updateStatusAndDetail(EventPublishingStatus.FAILED, "internal error");
+            throw new EventPublishingException("Error publishing message to kafka", e);
+        } catch (final RuntimeException e) {
+            item.updateStatusAndDetail(EventPublishingStatus.FAILED, "internal error");
+            throw new EventPublishingException("Error publishing message to kafka", e);
+        }
+    }
 
-            try {
-                kafkaProducer.send(record, kafkaSendCallback(item, done));
```
Comment (adyach) [inline src/main/java/org/zalando/nakadi/repository/kafka/KafkaTopicRepository.java]: Unused method in the code detected `kafkaSendCallback`

File: src/main/java/org/zalando/nakadi/repository/kafka/KafkaTopicRepository.java [hunk 8]
```diff
@@ -164,43 +165,89 @@ public boolean partitionExists(final String topic, final String partition) throw
...
-            } catch (InterruptException | SerializationException | BufferExhaustedException e) {
-                item.updateStatusAndDetail(EventPublishingStatus.FAILED, "internal error");
-                throw new EventPublishingException("Error publishing message to kafka", e);
-            }
+    private static boolean isExceptionShouldLeadToReset(
+            @Nullable final Exception exception) {
```
Comment (adyach) [inline src/main/java/org/zalando/nakadi/repository/kafka/KafkaTopicRepository.java]: Why not on the same line?

File: src/main/java/org/zalando/nakadi/repository/kafka/KafkaTopicRepository.java [hunk 9]
```diff
@@ -164,43 +165,89 @@ public boolean partitionExists(final String topic, final String partition) throw
...
+                    item.updateStatusAndDetail(EventPublishingStatus.FAILED, "internal error");
+                    result.complete(exception);
+                } else {
+                    item.setStep(EventPublishingStep.PUBLISHED);
+                    item.updateStatusAndDetail(EventPublishingStatus.SUBMITTED, "");
+                    result.complete(null);
+                }
+            }));
+            return result;
+        } catch (final InterruptException e) {
```
Comment (adyach) [inline src/main/java/org/zalando/nakadi/repository/kafka/KafkaTopicRepository.java]: Could you tell me who throws the exception in this method?

File: src/main/java/org/zalando/nakadi/domain/BatchItem.java (hunk 2 above)
Comment (antban) [inline src/main/java/org/zalando/nakadi/domain/BatchItem.java]: @adyach Because it was the only `this` usage

File: src/main/java/org/zalando/nakadi/repository/kafka/KafkaFactory.java (hunk 4 above)
Comment (antban) [inline src/main/java/org/zalando/nakadi/repository/kafka/KafkaFactory.java]: No, because of several reasons - Producers are taken and released under read lock, which means, that they are accessed from multiple threads simultaneously. That means, that it is impossible to use useCount variable without additional locking (that is why AtomicInteger)
`put` call to useCount map by itself is made with writeLock, while getting data is made without lock at all, and removing from this map is made with readLock. That means, that map by itself should be somehow synchronized (that is why `ConcurrentHashMap`)

File: src/main/java/org/zalando/nakadi/repository/kafka/KafkaTopicRepository.java (hunk 9 above)
Comment (antban) [inline src/main/java/org/zalando/nakadi/repository/kafka/KafkaTopicRepository.java]: `multifuture.get()`

File: src/main/java/org/zalando/nakadi/repository/kafka/KafkaFactory.java (hunk 5 above)
Comment (antban) [inline src/main/java/org/zalando/nakadi/repository/kafka/KafkaFactory.java]: Added javadocs: https://github.com/zalando/nakadi/pull/443/commits/a19fe4943cad21e3c7b8b9fa2fe506cee4428766

File: src/main/java/org/zalando/nakadi/repository/kafka/KafkaFactory.java (hunk 1 above)
Comment (antban) [inline src/main/java/org/zalando/nakadi/repository/kafka/KafkaFactory.java]: Added logging https://github.com/zalando/nakadi/pull/443/commits/a19fe4943cad21e3c7b8b9fa2fe506cee4428766

File: src/main/java/org/zalando/nakadi/repository/kafka/KafkaTopicRepository.java (hunk 6 above)
Comment (antban) [inline src/main/java/org/zalando/nakadi/repository/kafka/KafkaTopicRepository.java]: Well, for me this step is too important to hide it somewhere in call chain

File: src/main/java/org/zalando/nakadi/repository/kafka/KafkaTopicRepository.java (hunk 7 above)
Comment (antban) [inline src/main/java/org/zalando/nakadi/repository/kafka/KafkaTopicRepository.java]: Fixed https://github.com/zalando/nakadi/pull/443/commits/a19fe4943cad21e3c7b8b9fa2fe506cee4428766

File: src/main/java/org/zalando/nakadi/repository/kafka/KafkaTopicRepository.java (hunk 8 above)
Comment (antban) [inline src/main/java/org/zalando/nakadi/repository/kafka/KafkaTopicRepository.java]: Fixed: https://github.com/zalando/nakadi/pull/443/commits/a19fe4943cad21e3c7b8b9fa2fe506cee4428766

File: src/main/java/org/zalando/nakadi/domain/BatchItem.java (hunk 2 above)
Comment (adyach) [inline src/main/java/org/zalando/nakadi/domain/BatchItem.java]: @antban I did not get it. you still use `this` in `updateStatusAndDetail`.

File: src/main/java/org/zalando/nakadi/domain/EventPublishingStep.java [hunk 10]
```diff
@@ -1,5 +1,11 @@
...
 
 public enum EventPublishingStep {
-    NONE, VALIDATING, ENRICHING, PARTITIONING, PUBLISHING
+    NONE,
+    VALIDATING,
+    ENRICHING,
+    PARTITIONING,
+    PUBLISHING,
+    PUBLISHED,
+    FAILED_PUBLISH
```
Comment (v-stepanov) [inline src/main/java/org/zalando/nakadi/domain/EventPublishingStep.java]: We can't change this. This can break a lot of clients which expect only the specific values. This will be backward-incompatible change.
https://github.com/zalando/nakadi/blob/master/api/nakadi-event-bus-api.yaml#L1597

File: src/main/java/org/zalando/nakadi/repository/kafka/KafkaFactory.java [hunk 11]
```diff
@@ -1,29 +1,132 @@
...
+                useCount.get(activeProducer).incrementAndGet();
+                return activeProducer;
+            } else if (canCreate) {
+                activeProducer = new KafkaProducer<>(kafkaLocationManager.getKafkaProducerProperties());
+                useCountMetric.inc();
+                useCount.put(activeProducer, new AtomicInteger(1));
+                LOG.info("New producer instance created: " + activeProducer);
+                return activeProducer;
+            } else {
+                return null;
```
Comment (v-stepanov) [inline src/main/java/org/zalando/nakadi/repository/kafka/KafkaFactory.java]: I don't understand this branch. Why just not to create it in this case? Because this is exactly what you do outside of this method. You call takeUnderLock(false) and if it returns null you again call takeUnderLock(true).

File: src/main/java/org/zalando/nakadi/domain/BatchItem.java (hunk 2 above)
Comment (antban) [inline src/main/java/org/zalando/nakadi/domain/BatchItem.java]: Fixed https://github.com/zalando/nakadi/pull/443/commits/f793a564f0541cba936210948d5d2707cb1e474b

File: src/main/java/org/zalando/nakadi/repository/kafka/KafkaFactory.java (hunk 5 above)
Comment (antban) [inline src/main/java/org/zalando/nakadi/repository/kafka/KafkaFactory.java]: And fixed implementation: https://github.com/zalando/nakadi/pull/443/commits/f793a564f0541cba936210948d5d2707cb1e474b

File: src/main/java/org/zalando/nakadi/repository/kafka/KafkaTopicRepository.java (hunk 9 above)
Comment (antban) [inline src/main/java/org/zalando/nakadi/repository/kafka/KafkaTopicRepository.java]: ok, it is not there, it is Exception from kafka, which they rethrow without interrupting current thread back.

File: src/main/java/org/zalando/nakadi/repository/kafka/KafkaFactory.java (hunk 11 above)
Comment (antban) [inline src/main/java/org/zalando/nakadi/repository/kafka/KafkaFactory.java]: Because in first case it takes readLock (e.g. many threads can call this method at the same time). Second call is under write lock, which means, that only one thread is creating this producer. Thus, I guarantee, that I will now spawn several producers at the same time

File: src/main/java/org/zalando/nakadi/domain/EventPublishingStep.java (hunk 10 above)
Comment (antban) [inline src/main/java/org/zalando/nakadi/domain/EventPublishingStep.java]: Wow! Fixed https://github.com/zalando/nakadi/pull/443/commits/f7f405d972ece4b7cd56a77f56df44a08ee3007e

**Comment by antban** [comment]
@v-stepanov Tests are added, please, revote :)

File: src/main/java/org/zalando/nakadi/domain/BatchItem.java [hunk 12]
```diff
@@ -29,19 +28,23 @@ public void setPartition(final String partition) {
...
+    }
+
+    public EventPublishingStep getStep() {
+        return response.getStep();
     }
 
-    public synchronized void updateStatusAndDetail(final EventPublishingStatus publishingStatus, final String detail) {
-        this.response.setPublishingStatus(publishingStatus);
-        this.response.setDetail(detail);
+    public void updateStatusAndDetail(final EventPublishingStatus publishingStatus, final String detail) {
```
Comment (rcillo) [inline src/main/java/org/zalando/nakadi/domain/BatchItem.java]: There is no need for this combined update status and details anymore. It used to be like that due to a race condition protected by synchronized.

File: src/main/java/org/zalando/nakadi/domain/EventPublishingStep.java [hunk 13]
```diff
@@ -1,5 +1,9 @@
 package org.zalando.nakadi.domain;
 
 public enum EventPublishingStep {
-    NONE, VALIDATING, ENRICHING, PARTITIONING, PUBLISHING
+    NONE,
+    VALIDATING,
+    ENRICHING,
+    PARTITIONING,
+    PUBLISHING,
```
Comment (adyach) [inline src/main/java/org/zalando/nakadi/domain/EventPublishingStep.java]: ,

**Comment by adyach** [comment]
For me it looks good 👍

File: src/main/java/org/zalando/nakadi/repository/kafka/KafkaTopicRepository.java [hunk 14]
```diff
@@ -164,43 +164,89 @@ public boolean partitionExists(final String topic, final String partition) throw
...
+
+            // Now lets check for errors
+            final Optional<Exception> needReset = sendFutures.entrySet().stream()
+                    .filter(entry -> isExceptionShouldLeadToReset(entry.getValue().getNow(null)))
+                    .map(entry -> entry.getValue().getNow(null))
+                    .findAny();
+            if (needReset.isPresent()) {
+                LOG.info("Terminating producer while publishing to topic " + topicId +
+                        " because of unrecoverable exception", needReset.get());
+                kafkaFactory.terminateProducer(producer);
```
Comment (v-stepanov) [inline src/main/java/org/zalando/nakadi/repository/kafka/KafkaTopicRepository.java]: Wasn't the initial plan to retry publishing after the producer is recreated?

**Comment by v-stepanov** [comment]
:shipit:
